### PR TITLE
cleanup: Remove `DefIndexAddressSpace`

### PR DESCRIPTION
**Pull request by petrochenkov**
The scheme with two address spaces for `DefIndex` was needed in the past, but apparently not needed anymore (after removing `DefId`s from locals and `HirId`-ification).

### CODE REVIEW

**Comment by petrochenkov** [comment]
@Mark-Simulacrum 
I was waiting until local testing completes, the `try` is probably going to fail - some tests are not updated yet.

**Comment by Mark-Simulacrum** [comment]
Try doesn't require tests to pass (it's purely a dist build), so we should be fine if things build.

**Comment by michaelwoerister** [comment]
@rust-timer build fe050c93af5fe0a7417908fd5673bff06e0d2cd9

**Comment by michaelwoerister** [comment]
Thank you, @petrochenkov! Good riddance `:)`

**Comment by michaelwoerister** [comment]
Note, this could probably further simplified by using `newtype_index!` for `DefIndex` and switching things to `IndexVec<DefIndex, _>` where appropriate. But let's land this first `:)`

File: src/librustc/hir/map/definitions.rs [hunk 1]
```diff
@@ -26,116 +25,85 @@ use crate::util::nodemap::NodeMap;
...
-        let def_path_hashes_hi: Vec<DefPathHash> = Decodable::decode(d)?;
-
-        let index_to_key = [index_to_key_lo, index_to_key_hi];
-        let def_path_hashes = [def_path_hashes_lo, def_path_hashes_hi];
-
         Ok(DefPathTable {
-            index_to_key,
-            def_path_hashes,
+            index_to_key: Decodable::decode(d)?,
+            def_path_hashes : Decodable::decode(d)?,
```
Comment (eddyb) [inline src/librustc/hir/map/definitions.rs]: Doesn't this mean this `impl` can now be auto-derived? Same for encoding.

File: src/librustc_metadata/index.rs [hunk 2]
```diff
@@ -111,13 +109,10 @@ impl Index {
...
 
         // First we write the length of the lower range ...
-        buf.emit_raw_bytes(&(self.positions[0].len() as u32 / 4).to_le_bytes());
-        // ... then the values in the lower range ...
-        buf.emit_raw_bytes(&self.positions[0]);
-        // ... then the values in the higher range.
-        buf.emit_raw_bytes(&self.positions[1]);
-        LazySeq::with_position_and_length(pos as usize,
-            (self.positions[0].len() + self.positions[1].len()) / 4 + 1)
+        buf.emit_raw_bytes(&(self.positions.len() as u32 / 4).to_le_bytes());
```
Comment (eddyb) [inline src/librustc_metadata/index.rs]: This is unnecessary, since it indicated the split between lo/hi, which doesn't exist anymore.

**Comment by michaelwoerister** [comment]
@eddyb Yes, there's definitely further clean up possible now.

**Comment by eddyb** [comment]
@michaelwoerister I'll likely do it during the rebase of https://github.com/rust-lang/rust/pull/59953.

**Pull request merged**